File: python/115.py
"""Project Euler Problem 115.

Count tilings of a row of length n where blocks of minimum length m are
separated by at least one black square. Find the smallest n such that
F(m, n) exceeds one million for m = 50.
"""

from typing import List

# Constants for the problem
MIN_RED_LENGTH = 50
TARGET_WAYS = 1_000_000


def count_arrangements(min_length: int, length: int, cache: List[int]) -> int:
    """Count arrangements using dynamic programming."""
    while len(cache) <= length:
        idx = len(cache)

        if idx < min_length:
            cache.append(1)
            continue

        total = 1  # all-black configuration

        for block_length in range(min_length, idx + 1):
            max_start = idx - block_length
            for start in range(max_start + 1):
                remaining = idx - start - block_length
                if remaining == 0:
                    total += 1
                else:
                    total += cache[remaining - 1]

        cache.append(total)

    return cache[length]


def find_smallest_n(min_length: int, target: int) -> int:
    """Find the smallest n such that F(min_length, n) > target."""
    cache: List[int] = [1]
    n = min_length

    while True:
        total = count_arrangements(min_length, n, cache)
        if total > target:
            return n
        n += 1


def main() -> int:
    """Main function."""
    return find_smallest_n(MIN_RED_LENGTH, TARGET_WAYS)


if __name__ == "__main__":
    print(main())

File: c/115.c
/* Project Euler Problem 115: Counting Block Combinations II */
#include <stdio.h>

#define MAXN 500

static long long dp[MAXN];

static long long count_arrangements(int min_len, int length) {
    /* dp[i] = number of ways to tile row of length i */
    dp[0] = 1;
    for (int i = 1; i <= length; i++) {
        long long total = dp[i - 1]; /* grey square at end */
        for (int bl = min_len; bl <= i; bl++) {
            /* Place red block of length bl ending at position i */
            int remaining = i - bl;
            if (remaining == 0)
                total += 1;
            else if (remaining >= 1)
                total += dp[remaining - 1]; /* must have grey separator */
        }
        dp[i] = total;
    }
    return dp[length];
}

int main(void) {
    int min_len = 50;
    long long target = 1000000;

    for (int n = min_len; n < MAXN; n++) {
        long long val = count_arrangements(min_len, n);
        if (val > target) {
            printf("%d\n", n);
            return 0;
        }
    }
    return 1;
}
